### PR TITLE
Fix biginteger bitwise operations sign

### PR DESCRIPTION
**Pull request by fullkomnun**
The sign of a bitwise xor operation's result was not calculated correctly. The sign of the result was set the sign of the first operand which resulted in 

BigInteger.ONE xor BigInteger.ZERO == BigInteger.ONE while BigInteger.ZERO xor BigInteger.ONE == BigInteger.ZERO
so xor is not commutative and the sign of the result depends on the order of operands.

I amended the logic of the xor output's sign so that:
1. sign is Sign.NEGATIVE iff exactly one of the operands is negative (same as Java's BigInteger)
2. Sign is Sign.ZERO iff magnitude of the result has value of 0
3. else, sign is Sign.Positive

I also added a  basic test for xor of a non-zero number with zero that makes sure it is commutative.

I added an assertion that guards against initialisation of BigInteger when sign == Sign.ZERO but magnitude does not have a value of zero or when magnitude does have a value of 0 but sign != Sign.ZERO.

NOTE:
There could be more tests (maybe using generated data) to verify bitwise operators 
and sign of outputs of 'or' and 'and' needs to be verified as well.

### CODE REVIEW

**Comment by ionspin** [comment]
Hi Or, 

thank you for the pull request, it is really appreciated! I do have some comments that would need to be addressed first, especially since GitLab runners don't execute pipelines on pull requests from forked repos.

File: bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt [hunk 1]
```diff
@@ -47,6 +47,12 @@ class BigInteger internal constructor(wordArray: WordArray, val sign: Sign) : Bi
     constructor(short: Short) : this(arithmetic.fromShort(short), determinSignFromNumber(short))
     constructor(byte: Byte) : this(arithmetic.fromByte(byte), determinSignFromNumber(byte))
 
+    init {
```
Comment (ionspin) [inline bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt:50]: I agree that BigIntegers with zero magnitude should have Sign.ZERO, but adding this breaks a lot of things, just from one test run there were 12 failing tests.

 One way to fix this is to check wherever the code calls the constructor with sign parameter and check if the magnitude is zero and then correct the sign. 

Another way which should be easier, would be to change the constructor to something like:
`internal constructor(wordArray: WordArray, requestedSign: Sign)`
and then check the magnitude before setting the correct sign, i.e.:
```
    internal val sign: Sign = if (isResultZero(magnitude)) {
        Sign.ZERO
    } else {
        requestedSign
    }
```

**Comment by fullkomnun** [comment]
> Hi Or,
> 
> thank you for the pull request, it is really appreciated! I do have some comments that would need to be addressed first, especially since GitLab runners don't execute pipelines on pull requests from forked repos.

Sure

**Review by ionspin — CHANGES_REQUESTED** [review]

File: bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/integer/integer/BigIntegerBitwiseOperations.kt [hunk 2]
```diff
@@ -17,9 +17,27 @@
...
 /**
  * Created by Ugljesa Jovanovic
  * ugljesa.jovanovic@ionspin.com
  * on 01-Nov-2019
  */
-class BigIntegerBitwiseOperations
+class BigIntegerBitwiseOperations {
+
+    @Test
+    fun `xor with zero`() {
```
[the remark this answers is not in the document]
Comment (fullkomnun) [reply bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/integer/integer/BigIntegerBitwiseOperations.kt:32]: I like the ability to use ticks and let the name of the test describe it, to keep doing that we can either:
1. Use '_'(underscore) as a delimiter as done in 'BigIntegerReadmeTest'
2. Keep the method name and annotate it with @JsName("xor_with_zero").
Which one do you prefer?

[the remark this answers is not in the document]
Comment (ionspin) [reply bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/integer/integer/BigIntegerBitwiseOperations.kt:32]: It's not really strict but the rest of the library uses camel case, so go with that please, so in this case `xorWithZero`

[the remark this answers is not in the document]
Comment (fullkomnun) [reply bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/integer/integer/BigIntegerBitwiseOperations.kt]: Done

File: bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt (hunk 1 above)
Comment (fullkomnun) [reply bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt:50]: That is an interesting idea, but still allows initialising BigInteger with Sign.ZERO and a non-zero magnitude
so i kept the assertion for that case. All tests have passed.

Comment (fullkomnun) [reply bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt:50]: So to sum-up:
BigInteger(<zero magnitude>, <non-zero sign>) -> sign will be fixed to zero
BigInteger(<non-zero magnitude>, Sign.ZERO) -> assertion throws an error

Comment (ionspin) [reply bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/integer/BigInteger.kt:50]: Good point!

**Review by ionspin — APPROVED** [review]

**Comment by ionspin** [comment]
Looks good, once again thanks for the contribution, it's greatly appreciated!

I'm going to push your commits to a new request so that GitLab CI can run the build on Mac and Windows, and I'm going to add just a small commit with a newline at the end of the test you added because spotless is complaining about it. I'll reference the new pull request here once I create it, and then I'll close this pull request

**Comment by ionspin** [comment]
#143 with your commits is merged to master, in an hour or so once the build finished a new snapshot version will be released. I'll release a full version start of February, just to see if any other changes get in master by then.

If you want to use the snapshot version you'll have to add this:
```
repositories {
    maven {
        url = uri("https://oss.sonatype.org/content/repositories/snapshots")
    }
}
implementation("com.ionspin.kotlin:bignum:0.2.4-SNAPSHOT ")
```
to your build.gradle.kts

Once again, thanks for contributing!
